### PR TITLE
Ceren dev snaphot

### PR DESCRIPTION
**Pull request by CerenPaja**
for review

### CODE REVIEW

File: RunDNAm_small.sh
Comment (dn070017) [inline RunDNAm_small.sh:1]: [MUST] the scripts related to HPC can be put into maybe additional directory like scripts or scripts/sbatch

File: cavachon/utils/tensor_utils.py [hunk 1]
```diff
@@ -1,305 +1,305 @@
...
+
+        return gradients
+
+    @staticmethod
+    def create_backbone_layers(
+        n_layers: int = 2, #3
+        base_n_neurons: int = 32, #128
+        max_n_neurons: int = 64,  #2048
+        rate: int = 2,
+        activation: str = "swish", #"linear"
```
Comment (dn070017) [inline cavachon/utils/tensor_utils.py:110]: [NITS] I think it is safe to remove the old parameters in the comments (or we can keep them just for the ease of development)

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 2]
```diff
@@ -1,338 +1,901 @@
...
+from cavachon.distributions.mixture_multivariate_normal_diag_distribution import (
+    MixtureMultivariateNormalDiagDistribution,
+)
+from cavachon.distributions.multivariate_normal_diag_distribution import (
+    MultivariateNormalDiagDistribution,
+)
+from cavachon.environment.constants import Constants
+
+
+class PeriodicTSNECallback(tf.keras.callbacks.Callback):
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py:23]: [IMO] I think we may need a callback directory to put this (to keep things modular)

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 3]
```diff
@@ -1,338 +1,901 @@
...
+
+                        optimizer = tf.keras.optimizers.get(self.optimizer).__class__(
+                            learning_rate=learning_rate
+                        )
+                        self.model.compile(
+                            use_vanilla_kl=False,
+                            use_both_kl=True,
+                            optimizer=optimizer,
+                            loss_weights=loss_weights,
+                        )
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py:680]: [MUST] For the first part (parent to child), I thought we should set use_vanilla_kl=True, use_both_kl=False (note that in each iteration in the for-loop, we focus on the child component)

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 4]
```diff
@@ -1,338 +1,901 @@
...
+
+                        # Set initial KL weights
+                        self.model._vanilla_kl_weights[parent_name].assign(0.0)
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py]: [ASK] This sets the weight properly, but I'm not sure if there's any conflict with the self.model.compile configuration,

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 5]
```diff
@@ -1,338 +1,901 @@
...
+                        self.model._vanilla_kl_weights[component_name].assign(3.0)
+                        self.model._gmm_kl_weights[component_name].assign(0.0)
+
+                        kwargs_progressive = deepcopy(kwargs)
+                        kwargs_progressive.pop("epochs", None)
+                        callbacks_progressive = deepcopy(kwargs.get("callbacks", []))
+
+                        callbacks_progressive.append(
+                            ComponentWeightTransferCallback(
+                                total_epochs=n_progressive_epochs,
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py:697]: [NITS] In the future we may have to set total_epochs to n_progressive_epochs // 2, and next phase n_progressive_epochs -  n_progressive_epochs // 2 (just write it down for the record)

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 6]
```diff
@@ -1,338 +1,901 @@
...
+        # Check if single component (for PeriodicTSNECallback decision)
+        is_single_component = len(self.training_order) == 1
+
+        for component_order, train_components in enumerate(self.training_order):
+            component_name = train_components[0]
+
+            # PROGRESSIVE SECTION - Phase 2 (Child Only)
+
+            if self.run_progressive_training.get(component_name):
+                parent_name = self._get_parent_component(component_name)
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py]: [MUST] Parent name may be more than one (one child can have more than one parent), so we may have to tranverse the list (or set), and set the weight for all parents first, and then set the child.
(See commments above)

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 7]
```diff
@@ -1,338 +1,901 @@
...
+        self.model.trainable = True
+
+    def _get_parent_component(self, child_name: str) -> Optional[str]:
+        """Get parent component name for the child."""
+        for component_config in self.component_configs:
+            if component_config.get("name") == child_name:
+                conditioned_on_z_hat = component_config.get(
+                    Constants.CONFIG_FIELD_COMPONENT_CONDITION_Z_HAT, []
+                )
+                if len(conditioned_on_z_hat) > 0:
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py:600]: [MUST] Shall return a List or Set

**Review by dn070017 — CHANGES_REQUESTED** [review]
Hi Ceren, here's some of my comments, you can first check the MUST tag and tries to fix some of the issue (or we can discuss if other features needs to be implemented, or we can ignore those things first), for NITS or ASK or IMO, you don't necessarily need to change anything, just some of my thought on the code.

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 8]
```diff
@@ -1,338 +1,901 @@
...
+                    return conditioned_on_z_hat[0]
+        return None
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py]: return List[str] # keep zero-length list if no parent

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 9]
```diff
@@ -1,338 +1,901 @@
...
+
+        for component_order, train_components in enumerate(self.training_order):
+            component_name = train_components[0]
+
+            # PROGRESSIVE SECTION - Phase 2 (Child Only)
+
+            if self.run_progressive_training.get(component_name):
+                parent_name = self._get_parent_component(component_name)
+
+                if parent_name is not None:
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py]: for parent_name in parent_name_list

File: cavachon/scheduler/sequential_training_scheduler.py [hunk 10]
```diff
@@ -1,338 +1,901 @@
...
+                            use_vanilla_kl=False,
+                            use_both_kl=True,
+                            optimizer=optimizer,
+                            loss_weights=loss_weights,
+                        )
+
+                        # Set initial KL weights
+                        self.model._vanilla_kl_weights[parent_name].assign(0.0)
+                        self.model._gmm_kl_weights[parent_name].assign(1.0)
+                        self.model._vanilla_kl_weights[component_name].assign(3.0)
```
Comment (dn070017) [inline cavachon/scheduler/sequential_training_scheduler.py:689]: this needs to be outside the parent loop

**Review by dn070017 — APPROVED** [review]
LGTM, just a side note that originally CAVACHON supports conditioned on z, the code implemented now only support conditioned on z_hat (we can fix this later as most of the user will probably just conditioned on z_hat)
Note that I only review the code statically, please run some tests before merging to develop branch

**Comment by dn070017** [comment]
remember to use squash merge (you can select from the expand error on the right of merge pull request)

File: cavachon/workflow/workflow.py [hunk 11]
```diff
@@ -382,6 +388,199 @@ def visualize_embedding(self) -> None:
...
+        for component_config in self.config.components:
+            component_name = component_config.get("name")
+            parent_names = component_config.get("conditioned_on_z_hat", [])
+
+            if not parent_names:
+                continue  # Skip non-hierarchical components
+
+            # Find modality for saving results
+            modality_names = component_config.get("modality_names", [])
+            modality = modality_names[0] if modality_names else component_name
```
Comment (dn070017) [inline cavachon/workflow/workflow.py]: [MUST] Ideally we should allow for the case that one component output multiple modalities (put a comment which contains TODO so we can come back to unfix features in the future quickly)

File: cavachon/workflow/workflow.py [hunk 12]
```diff
@@ -382,6 +388,199 @@ def visualize_embedding(self) -> None:
...
+                component_name
+            ].hierarchical_encoder.b_network.get_weights()
+
+            # Split W by parent dimensions
+            W_parts = []
+            offset = 0
+            for dim in parent_dims:
+                W_parts.append(W[offset : offset + dim, :])
+                offset += dim
+            W_child = W[offset:, :]
```
Comment (dn070017) [inline cavachon/workflow/workflow.py]: [MUST] W_child is scaled using r_networks. Check https://github.com/dn070017/CAVACHON/blob/develop/cavachon/modules/base/hierarchical_encoder.py
So basically, the psedo-code is like
b_networks([z_parent_1, z_parent_2..., r_networks(z_child)])
So you might have to take r_networks parameters into count for z_child.
Note that originally there's also bias in the r_networks and b_networks, maybe you can disable it to make sure the approximation works. Just add use_bias=False to the dense layer defined in https://github.com/dn070017/CAVACHON/blob/develop/cavachon/modules/base/hierarchical_encoder.py#L58

**Review by dn070017 — CHANGES_REQUESTED** [review]
Please check some of my comments, and make sure the r_networks is fixed (we might also need to fix the bias)
For other comments labeled as MUST, but you don't want to fix it now, you can add a `# TODO: explain what we should do next`
comment so we can quickly refer to them in the future
